README.md
# What is USD for Maya?

USD for Maya is a project to create a Maya plugin, as well as reusable libraries, that provide translation and editing capabilities for Pixar Animation Studios Universal Scene Description (USD).

Read more about Pixar's USD [here](http://openusd.org)


## Motivation
*Why yet another Maya plugin?*

As USD gains in popularity many studios have been wondering what plugin they should deploy within their pipelines.  Two popular choices, used individually as well as together, have been the Pixar USDMaya plugin that has been part of USD itself, and Animal Logic's, which was separately released as Open Source.  Both plugins have a strong community of users, but some have questioned when Autodesk would eventually release an officially supported plugin themselves.  The goal of this project is to work directly with Pixar and Animal Logic, as well as other key contributors, in order to merge the best of both into a single supported plugin.  

Pixar maintains a list of contributors to USD in their documentation [here](https://graphics.pixar.com/usd/docs/USD-Contributors.html), and as we start working together to merge the two existing plugins it is also important to acknowledge the amount of work that individuals put in on the Animal Logic plugin as well.  The list of those contributors is kept [here](doc/AL_CONTRIBUTORS.md).

## Getting Help
Both Animal Logic and Pixar have their own Google interest groups which will remain great places to go with questions.

Pixar forum is [here](https://groups.google.com/forum/#!forum/usd-interest)  
Animal Logic forum is [here](https://groups.google.com/forum/#!forum/al_usdmaya-discussion)


## Detailed Documentation

+ [Contributing](doc/CONTRIBUTING.md)
+ [Building](doc/build.md)
+ [License](doc/LICENSE.md)
+ [Developer](doc/DEVELOPER.md)


CHANGELOG.md
## v0.34.3 (2019-09-11)

### Added
+ Use new MPxTransform_BoundingBox class in Maya2019 update 2 (and successive 2019 versions) to allow framing of objects under Custom Transforms
+ File Import - a prim path ("Prim Path") argument to allow import from a specific prim

### Fixed
+ AL_usdmaya_TranslatePrim with the recursive flag turned on will not complain about non-updateable prims anymore.
+ Translation of colourSets
+ uuid comparision when plug names are identical - with Maya References, two nodes under different dag paths could share the same name including the namespace. Try to resolve to the nodes' uuids to avoid missing the animated attributes and fix
+ Unable to translate prims that have been selected.

## v0.34.2 (2019-08-29)

### Added
+ Track editTarget and dirty layers when opening from stageCacheId 

### Fixed
+ #155/OSS tests were using the string "USD_LIBRARY_DIR" instead of the variable
+ #154/OSS Methods to query information about renderer plugins are now static
+ #153/OSS fix boolean return value of DirectionalLight::updateUsdPrim
+ #152/OSS if renderGuides are shown, they should be selectable ToTest
+ OSS Linux and Windows Build fixes (DSO Exports, CMake, etc)



## v0.34.1 (2019-08-21)

### Changed

+ Visibility state is now applied to transforms during mesh import.

## v0.34.0 (2019-08-19)

### Added

+ Introduce AL_USDTransaction module, see [docs](usdtransaction/AL/usd/transaction/README.md)

## v0.33.5 (2019-08-19)
### Fixed

+ Potential crash when serialising internal transform references

## v0.33.4 (2019-08-16)
### Added

+ Add tests to cover the editTarget and session layer serialisation and deserialisation issues

### Changed
+ During serialisation do not store session layer name on proxyshape node if it is clean 
+ Make layer deserialised by layerManager reloadable and simplify the lengthy layer debugging message 

## v0.33.3 (2019-08-14)
### Fixed
+ Setting maya rotation values via attributes causes nonsense values to appear only when using ZXY rotation orders. Use an MFnTransform instead

## v0.33.2 (2019-08-12)
### Added
+ added option var to control pushToPrim 
+ use the pushToPrim option var within the translate prim command

## v0.33.0 (2019-08-12)
### Added 
+ Support for Python Translators
+ TranslatorBase/TranslatorContext python bindings
+ Register Python Translators against AssetType Metadata (#915)
+ Allow Filtering of python plugins by registration type (#918)
+ python translator unittests (#923)

## v0.32.39 (2019-08-14)
### Fixed
+ Changed way that rotation values were being set on the custom transform nodes to work around a maya bug

## v0.32.38 (2019-08-12)
### Changed
+ pushToPrim flag default now controlled by AL_usdmaya_pushToPrim primvar

## v0.32.37 (2019-08-09)
### Added
+ optionVar and menu item to control the default state of pushToPrim

## v0.32.36 (2019-08-08)
### Fixed
+ Prevent pref data being displayed as a colour set
- Incorrect time value used when evaluating the proxy shape bounding box
- Bounding box cache not updated when transform values change

## v0.32.35 (2019-08-07)
### Fixed
+ Crashes when deleting nodes as a result of a variant switch
### Changed
+ Meshes with time bounds are treated as animated
+ Reduced the amount of work done under the hood when transform nodes are manipulated in the viewport.

## v0.32.34 (2019-08-05)
### Fixed
+ Regression that no longer reloaded the stage when filePath changes

## v0.32.33 (2019-08-02)
### Added
+ Ability to load proxy shape from the stage cache
### Fixed
+ various bugs within the plugin translator system

## v0.32.32 (2019-07-26)
### Fixed
+ import/export of vertex colours

## v0.32.31 (2019-07-12)
### Fixed
+ invalid array access when diffing mesh edge creases

## v0.32.30 (2019-07-05)
### Added
+ Menu item to control whether selection is enabled in the viewport (controlled via the AL_usdmaya_selectionEnabled optionVar)
### Fixed
+ constructLockPrims could end up in an infinite loop

## v0.32.29 (2019-07-03)
### Added 
+ AL_usdmaya_TranslatePrim can now recurse through a hierarchy of prims
+ AL_usdmaya_TranslatePrim can now update a set of prims
### Fixed
+ Failing animation export unit tests 

## v0.32.28 (2019-06-28)
### Changed
+ VP2 selection API now suppported in Maya2018 (20180600)
+ visibility flag now correctly handled by the mesh translator

## v0.32.27 (2019-06-06)
### Fixed
+ Potential crash when importing a scene 

## v0.32.26 (2019-06-05)
### Added
+ added ability to set the readAnimatedValues flag from the translate prim command
### Fixed
+ Ensure prims get correctly relocked when reactivating a prim
+ Refresh viewport when external changes are made to the stage

## v0.32.25 (2019-05-23)
### Changed
+ previously if a namespace was not provided for a maya reference, the reference would not be loaded. Now we
  generate a fallback namespace based on the prim path, so "/root/mayaRef" will use the namespace "root_mayaRef"

## v0.32.24 (2019-05-14)
### Fixed
+ Visible in reflections/refractions now enabled by default when translating a mesh

## v0.32.23 (2019-05-14)
### Fixed
+ Transform attributes can no longer change when locked.

## v0.32.22 (2019-05-07)
### Changed
+ Import mesh normals in unlocked state

## v0.32.21 (2019-05-07)
### Fixed
+ Selection highlighting visible in wireframe mode.

## v0.32.20 (2019-05-02)
### Added
+ Added "AL_usdmaya_readAnimatedValues" option var to control default setting of readAnimatedValues attribute

## v0.32.19 (2019-04-18)
### Added
+ Added "serializedArCtx" attribute as non-storable, to prevent errors when loading old files. 
### Fixed
+ Ensure readAnimatedValues is correctly set on transform creation.
+ Ensure transform values are correctly updated when scrubbing through the timeline.
+ Wireframe selection rendering trailed behind the geom by 1 frame when scrubbing throught the timeline

## v0.32.18 (2019-04-15)
### Fixed
+ Curve width export now supports widths that are MFn::kFloatArray or MFn::kDoubleArray


build.py
from distutils.spawn import find_executable

import argparse
import contextlib
import datetime
import distutils
import multiprocessing
import os
import platform
import re
import shlex
import shutil
import subprocess
import sys

############################################################
# Helpers for printing output
verbosity = 1

def Print(msg):
    if verbosity > 0:
        print msg

def PrintWarning(warning):
    if verbosity > 0:
        print "WARNING:", warning

def PrintStatus(status):
    if verbosity >= 1:
        print "STATUS:", status

def PrintInfo(info):
    if verbosity >= 2:
        print "INFO:", info

def PrintCommandOutput(output):
    if verbosity >= 3:
        sys.stdout.write(output)

def PrintError(error):
    if verbosity >= 3 and sys.exc_info()[1] is not None:
        import traceback
        traceback.print_exc()
    print "ERROR:", error

############################################################

def Windows():
    return platform.system() == "Windows"

def Linux():
    return platform.system() == "Linux"

def MacOS():
    return platform.system() == "Darwin"

def GetCommandOutput(command):
    """Executes the specified command and returns output or None."""
    try:
        return subprocess.check_output(
            shlex.split(command), stderr=subprocess.STDOUT).strip()
    except subprocess.CalledProcessError:
        pass
    return None

def GetXcodeDeveloperDirectory():
    """Returns the active developer directory as reported by 'xcode-select -p'.
    Returns None if none is set."""
    if not MacOS():
        return None

    return GetCommandOutput("xcode-select -p")

def GetVisualStudioCompilerAndVersion():
    """Returns a tuple containing the path to the Visual Studio compiler
    and a tuple for its version, e.g. (14, 0). If the compiler is not found
    or version number cannot be determined, returns None."""
    if not Windows():
        return None

    msvcCompiler = find_executable('cl')
    if msvcCompiler:
        # VisualStudioVersion environment variable should be set by the
        # Visual Studio Command Prompt.
        match = re.search(
            "(\d+).(\d+)",
            os.environ.get("VisualStudioVersion", ""))
        if match:
            return (msvcCompiler, tuple(int(v) for v in match.groups()))
    return None

def IsVisualStudio2017OrGreater():
    VISUAL_STUDIO_2017_VERSION = (15, 0)
    msvcCompilerAndVersion = GetVisualStudioCompilerAndVersion()
    if msvcCompilerAndVersion:
        _, version = msvcCompilerAndVersion
        return version >= VISUAL_STUDIO_2017_VERSION
    return False

def GetPythonInfo():
    """Returns a tuple containing the path to the Python executable, shared
    library, and include directory corresponding to the version of Python
    currently running. Returns None if any path could not be determined. This
    function always returns None on Windows or Linux.

    This function is primarily used to determine which version of
    Python USD should link against when multiple versions are installed.
    """
    # We just skip all this on Windows. Users on Windows are unlikely to have
    # multiple copies of the same version of Python, so the problem this
    # function is intended to solve doesn't arise on that platform.
    if Windows():
        return None

    # We also skip all this on Linux. The below code gets the wrong answer on
    # certain distributions like Ubuntu, which organizes libraries based on
    # multiarch. The below code yields /usr/lib/libpython2.7.so, but
    # the library is actually in /usr/lib/x86_64-linux-gnu. Since the problem
    # this function is intended to solve primarily occurs on macOS, so it's
    # simpler to just skip this for now.
    if Linux():
        return None

    try:
        import distutils.sysconfig

        pythonExecPath = None
        pythonLibPath = None

        pythonPrefix = distutils.sysconfig.PREFIX
        if pythonPrefix:
            pythonExecPath = os.path.join(pythonPrefix, 'bin', 'python')
            pythonLibPath = os.path.join(pythonPrefix, 'lib', 'libpython2.7.dylib')

        pythonIncludeDir = distutils.sysconfig.get_python_inc()
    except:
        return None

    if pythonExecPath and pythonIncludeDir and pythonLibPath:
        # Ensure that the paths are absolute, since depending on the version of
        # Python being run and the path used to invoke it, we may have gotten a
        # relative path from distutils.sysconfig.PREFIX.
        return (
            os.path.abspath(pythonExecPath),
            os.path.abspath(pythonLibPath),
            os.path.abspath(pythonIncludeDir))

    return None

def GetCPUCount():
    try:
        return multiprocessing.cpu_count()
    except NotImplementedError:
        return 1

def Run(cmd, logCommandOutput=True):
    """Run the specified command in a subprocess."""
    PrintInfo('Running "{cmd}"'.format(cmd=cmd))

    with open(context.logFileLocation, "a") as logfile:
        logfile.write(datetime.datetime.now().strftime("%Y-%m-%d %H:%M"))
        logfile.write("\n")
        logfile.write(cmd)
        logfile.write("\n")

        # Let exceptions escape from subprocess calls -- higher level
        # code will handle them.
        if logCommandOutput:
            p = subprocess.Popen(shlex.split(cmd), stdout=subprocess.PIPE,
                                 stderr=subprocess.STDOUT)
            while True:
                l = p.stdout.readline()
                if l != "":
                    logfile.write(l)
                    PrintCommandOutput(l)
                elif p.poll() is not None:
                    break
        else:
            p = subprocess.Popen(shlex.split(cmd))
            p.wait()

    if p.returncode != 0:
        # If verbosity >= 3, we'll have already been printing out command output
        # so no reason to print the log file again.
        if verbosity < 3:
            with open(context.logFileLocation, "r") as logfile:
                Print(logfile.read())
        raise RuntimeError("Failed to run '{cmd}'\nSee {log} for more details."
                           .format(cmd=cmd, log=os.path.abspath(context.logFileLocation)))

def BuildVariant(context): 
    if context.buildDebug:
        return "Debug"
    elif context.buildRelease:
        return "Release"
    elif context.buildRelWithDebug:
        return "RelWithDebInfo"
    return "RelWithDebInfo"

def FormatMultiProcs(numJobs, generator):
    tag = "-j"
    if generator:
        if "Visual Studio" in generator:
            tag = "/M:"
        elif "Xcode" in generator:
            tag = "-j "

    return "{tag}{procs}".format(tag=tag, procs=numJobs)

############################################################
# contextmanager
@contextlib.contextmanager
def CurrentWorkingDirectory(dir):
    """Context manager that sets the current working directory to the given
    directory and resets it to the original directory when closed."""
    curdir = os.getcwd()
    os.chdir(dir)
    try:
        yield
    finally:
        os.chdir(curdir)

############################################################
# CMAKE
def RunCMake(context, extraArgs=None, stages=None):
    """Invoke CMake to configure, build, and install a library whose 
    source code is located in the current working directory."""

    srcDir = os.getcwd()
    instDir = context.instDir
    buildDir = context.buildDir

    if 'clean' in stages and os.path.isdir(buildDir):
        shutil.rmtree(buildDir)

    if 'clean' in stages and os.path.isdir(instDir):
        shutil.rmtree(instDir)

    if not os.path.isdir(buildDir):
        os.makedirs(buildDir)

    generator = context.cmakeGenerator

    # On Windows, we need to explicitly specify the generator to ensure we're
    # building a 64-bit project. (Surely there is a better way to do this?)
    # TODO: figure out exactly what "vcvarsall.bat x64" sets to force x64
    if generator is None and Windows():
        if IsVisualStudio2017OrGreater():
            generator = "Visual Studio 15 2017 Win64"
        else:
            generator = "Visual Studio 14 2015 Win64"

    if generator is not None:
        generator = '-G "{gen}"'.format(gen=generator)

    # On MacOS, enable the use of @rpath for relocatable builds.
    osx_rpath = None
    if MacOS():
        osx_rpath = "-DCMAKE_MACOSX_RPATH=ON"

    # get build variant 
    variant= BuildVariant(context)

    with CurrentWorkingDirectory(buildDir):
        # recreate build_log.txt everytime the script runs
        if os.path.isfile(context.logFileLocation):
            os.remove(context.logFileLocation)

        if 'configure' in stages:
            Run('cmake '
                '-DCMAKE_INSTALL_PREFIX="{instDir}" '
                '-DCMAKE_BUILD_TYPE={variant} '
                '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON'
                '{osx_rpath} '
                '{generator} '
                '{extraArgs} '
                '"{srcDir}"'
                .format(instDir=instDir,
                        variant=variant,
                        srcDir=srcDir,
                        osx_rpath=(osx_rpath or ""),
                        generator=(generator or ""),
                        extraArgs=(" ".join(extraArgs) if extraArgs else "")))
 
        installArg = ""
        if 'install' in stages:
            installArg = "--target install"

        if 'build' in stages or 'install' in stages:
            Run("cmake --build . --config {variant} {installArg} -- {multiproc}"
                .format(variant=variant,
                        installArg=installArg,
                        multiproc=FormatMultiProcs(context.numJobs, generator)))

############################################################
# Maya USD
def InstallMayaUSD(context, buildArgs, stages):
    with CurrentWorkingDirectory(context.mayaUsdSrcDir):
        extraArgs = []
        stagesArgs = []
        if context.mayaLocation:
            extraArgs.append('-DMAYA_LOCATION="{mayaLocation}"'
                             .format(mayaLocation=context.mayaLocation))

        if context.pxrUsdLocation:
            extraArgs.append('-DPXR_USD_LOCATION="{pxrUsdLocation}"'
                             .format(pxrUsdLocation=context.pxrUsdLocation))

        if context.devkitLocation:
            extraArgs.append('-DMAYA_DEVKIT_LOCATION="{devkitLocation}"'
                             .format(devkitLocation=context.devkitLocation))

        # Add on any user-specified extra arguments.
        extraArgs += buildArgs

        # Add on any user-specified stages arguments.
        stagesArgs += stages

        RunCMake(context, extraArgs, stagesArgs)

############################################################
# ArgumentParser

parser = argparse.ArgumentParser(
    formatter_class=argparse.RawDescriptionHelpFormatter)

parser.add_argument("workspace_location", type=str,
                    help="Directory where the project use as a workspace to build and install plugin/libraries.")

parser.add_argument("--generator", type=str,
                    help=("CMake generator to use when building libraries with "
                          "cmake"))

parser.add_argument("-v", "--verbosity", type=int, default=verbosity,
                    help=("How much output to print while building: 0 = no "
                          "output; 1 = warnings + status; 2 = info; 3 = "
                          "command output and tracebacks (default: "
                          "%(default)s)"))

parser.add_argument("--build-location", type=str,
                    help=("Set Build directory "
                          "(default: <workspace_location>/build-location)"))

parser.add_argument("--install-location", type=str,
                    help=("Set Install directory "
                          "(default: <workspace_location>/install-location)"))

parser.add_argument("--maya-location", type=str,
                    help="Directory where Maya is installed.")

parser.add_argument("--pxrusd-location", type=str,
                    help="Directory where Pixar USD is installed.")

parser.add_argument("--devkit-location", type=str,
                    help="Directory where Maya Devkit is installed.")

parser.add_argument("--build-debug", dest="build_debug", action="store_true",
                    help="Build in Debug mode")

parser.add_argument("--build-release", dest="build_release", action="store_true",
                    help="Build in Release mode")

parser.add_argument("--build-relwithdebug", dest="build_relwithdebug", action="store_true",
                    help="Build in RelWithDebInfo mode")

parser.add_argument("--build-args", type=str, nargs="*", default=[],
                   help=("Comma-separated list of arguments passed into CMake when building libraries"))

parser.add_argument("--stages", type=str, nargs="*", default=['clean','configure','build','install'],
                   help=("Comma-separated list of stages to execute.( possible stages: clean, configure, build, install)"))

parser.add_argument("-j", "--jobs", type=int, default=GetCPUCount(),
                    help=("Number of build jobs to run in parallel. "
                          "(default: # of processors [{0}])"
                          .format(GetCPUCount())))

args = parser.parse_args()
verbosity = args.verbosity

############################################################
# InstallContext
class InstallContext:
    def __init__(self, args):

        # Assume the project's top level cmake is in the current source directory
        self.mayaUsdSrcDir = os.path.normpath(
            os.path.join(os.path.abspath(os.path.dirname(__file__))))

        # Build type
        # Must be done early, so we can call BuildVariant(self)
        self.buildDebug = args.build_debug
        self.buildRelease = args.build_release
        self.buildRelWithDebug = args.build_relwithdebug

        # Workspace directory 
        self.workspaceDir = os.path.abspath(args.workspace_location)

        # Build directory
        self.buildDir = (os.path.abspath(args.build_location) if args.build_location
                         else os.path.join(self.workspaceDir, "build", BuildVariant(self)))

        # Install directory
        self.instDir = (os.path.abspath(args.install_location) if args.install_location
                         else os.path.join(self.workspaceDir, "install", BuildVariant(self)))

        # CMake generator
        self.cmakeGenerator = args.generator

        # Number of jobs
        self.numJobs = args.jobs
        if self.numJobs <= 0:
            raise ValueError("Number of jobs must be greater than 0")

        # Maya Location
        self.mayaLocation = (os.path.abspath(args.maya_location)
                                if args.maya_location else None)

        # PXR USD Location
        self.pxrUsdLocation = (os.path.abspath(args.pxrusd_location)
                                if args.pxrusd_location else None)

        # Maya Devkit Location
        self.devkitLocation = (os.path.abspath(args.devkit_location)
                                if args.devkit_location else None)

        # Log File Name
        logFileName="build_log.txt"
        self.logFileLocation=os.path.join(self.buildDir, logFileName)

        # Build arguments
        self.buildArgs = list()
        for argList in args.build_args:
            for arg in argList.split(","):
                self.buildArgs.append(arg)

        # Stages arguments
        self.stagesArgs = list()
        for argList in args.stages:
            for arg in argList.split(","):
                self.stagesArgs.append(arg)
try:
    context = InstallContext(args)
except Exception as e:
    PrintError(str(e))
    sys.exit(1)

# Summarize
summaryMsg = """
Building with settings:
  Source directory          {mayaUsdSrcDir}
  Workspace directory       {workspaceDir}
  Build directory           {buildDir}
  Install directory         {instDir}
  Variant                   {buildVariant}
  CMake generator           {cmakeGenerator}
  Build Log                 {logFileLocation}"""

if context.buildArgs:
  summaryMsg += """
  Extra Build arguments     {buildArgs}"""

if context.stagesArgs:
  summaryMsg += """
  Stages arguments          {stagesArgs}"""

summaryMsg = summaryMsg.format(
    mayaUsdSrcDir=context.mayaUsdSrcDir,
    workspaceDir=context.workspaceDir,
    buildDir=context.buildDir,
    instDir=context.instDir,
    logFileLocation=context.logFileLocation,
    buildArgs=context.buildArgs,
    stagesArgs=context.stagesArgs,
    buildVariant=BuildVariant(context),
    cmakeGenerator=("Default" if not context.cmakeGenerator
                    else context.cmakeGenerator)
)

Print(summaryMsg)

# Install MayaUSD
InstallMayaUSD(context, context.buildArgs, context.stagesArgs)

# Ensure directory structure is created and is writable.
for dir in [context.workspaceDir, context.buildDir, context.instDir]:
    try:
        if os.path.isdir(dir):
            testFile = os.path.join(dir, "canwrite")
            open(testFile, "w").close()
            os.remove(testFile)
        else:
            os.makedirs(dir)
    except Exception as e:
        PrintError("Could not write to directory {dir}. Change permissions "
                   "or choose a different location to install to."
                   .format(dir=dir))
        sys.exit(1)

Print("""Success Maya USD build and install !!!!""")
